### PR TITLE
feat(typescript-react-apollo): Apollo Client `useFragment` hook

### PR DESCRIPTION
**Pull request by rickdunkin**
## Description

Please include a summary of the change and which issue is fixed. Please also include relevant
motivation and context. List any dependencies that are required for this change.

Related #482  

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Screenshots/Sandbox (if appropriate/relevant):

Adding links to sandbox or providing screenshots can help us understand more about this PR and take
action on it as appropriate

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can
reproduce. Please also list any relevant details for your test configuration

- [x] wrote a new test, updated a `dev-test` file, ran `yarn build`, `yarn generate:examples`, `yarn test` and all completed successfully
- [x] ran the updated code in my project at work and had `useFragment` hooks generated as expected

**Test Environment**:

- OS: macOS Sonoma 14.1
- `@graphql-codegen/typescript-react-apollo@4.1.0`:
- NodeJS: 20.9.0

## Checklist:

- [x] I have followed the
      [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the
      style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

## Further comments

The solution(s) I came up with here may not be the most optimal, this is the first time I've looked through the code used for `graphql-codegen` so there could be better ways to do this

### CODE REVIEW

File: packages/plugins/typescript/react-apollo/src/visitor.ts [hunk 1]
```diff
@@ -583,4 +589,58 @@ export class ReactApolloVisitor extends ClientSideBaseVisitor<
...
+
+      const operationTypeSuffix: string = this.getOperationSuffix(fragmentName, operationType);
+
+      const operationResultType: string = this.convertName(nodeName, {
+        suffix: operationTypeSuffix + this._parsedConfig.operationResultSuffix,
+      });
+
+      const IDType = this.scalars.ID ?? 'string';
+
+      const hook = `export function use${fragmentName}(id: ${IDType}) {
```
Comment (rohit-gohri) [inline packages/plugins/typescript/react-apollo/src/visitor.ts:626]: I'm also interested in this but my usecase involves custom keyFields, would be great if the input type can be inferred from imported type policies or be a generic object instead like:

```suggestion
      const hook = `export function use${fragmentName}<F = { id: ${IDType} }>(from: F) {
```

inferring from type policy can also throw type errors on invalid use of `useFragment`

Comment (rickdunkin) [reply packages/plugins/typescript/react-apollo/src/visitor.ts]: Added support for `keyFields`, thanks for the suggestion!

**Review by saihaj — APPROVED** [review]
Thank you for this and sorry for the delays

**Comment by saihaj** [comment]
Published [@graphql-codegen/typescript-react-apollo@4.2.0](https://github.com/dotansimha/graphql-code-generator-community/releases/tag/release-1706891918038)

**Pull request merged**